pkg/controller/wildflyserver/wildflyserver_controller.go

```diff
@@ -504,20 +550,339 @@ func (r *ReconcileWildFlyServer) routeForWildFly(w *wildflyv1alpha1.WildFlyServe
+	}
+	// Ensure the env variables are up to date
+	for _, env := range spec.Env {
+		if !matches(&statefuleSet.Spec.Template.Spec.Containers[0], env) {
+			log.Info("Updated statefulset env", "StatefulSet.Namespace", statefuleSet.Namespace, "StatefulSet.Name", statefuleSet.Name, "Env", env)
+			update = true
+		}
+	}
+	// Ensure the envFrom variables are up to date
+	envFrom := spec.EnvFrom
+	if !reflect.DeepEqual(statefuleSet.Spec.Template.Spec.Containers[0].EnvFrom, envFrom) {
+		log.Info("Updating envFrom", "StatefulSet.Namespace", statefuleSet.Namespace, "StatefulSet.Name", statefuleSet.Name)
+		statefuleSet.Spec.Template.Spec.Containers[0].EnvFrom = envFrom
+		update = true
+	}
+
+	return update
+}
+
+// TODO: operator should consider changing the recovery properties of narayana backoff period
+//    -Dcom.arjuna.ats.arjuna.common.RecoveryEnvironmentBean.recoveryBackoffPeriod=1 -Dcom.arjuna.ats.arjuna.common.RecoveryEnvironmentBean.periodicRecoveryPeriod=1 -Dcom.arjuna.ats.jta.common.JTAEnvironmentBean.orphanSafetyInterval=1
+func checkRecovery(reqLogger logr.Logger, scaleDownPod *corev1.Pod) (bool, string, error) {
+	scaleDownPodName := scaleDownPod.ObjectMeta.Name
+	scaleDownPodIP := scaleDownPod.Status.PodIP
+	jbossHome := getJbossHome()
+
+	// Reading timestamp for the latest log record
+	scaleDownPodLogTimestampAtStart, err := wildflyutil.ObtainLogLatestTimestamp(scaleDownPod)
+	if err != nil {
+		return false, "", fmt.Errorf("Failed to read log from scaling down pod '%v', error: %v", scaleDownPodName, err)
+	}
+
+	// TODO: consider if we can get rid of the recovery listener enablement to speed up the scale down processing
+	// Enabling recovery listener to speed up recovery processing
+	jsonResult, err := wildflyutil.ExecuteMgmtOp(scaleDownPod, jbossHome, wildflyutil.MgmtOpTxnEnableRecoveryListener)
+	if err != nil {
+		return false, "", fmt.Errorf("Cannot enable transaction recovery listener for scaling down pod %v, error: %v", scaleDownPodName, err)
+	}
+	if !wildflyutil.IsMgmtOutcomeSuccesful(jsonResult) {
+		return false, "", fmt.Errorf("Failed to enable transaction recovery listener for scaling down pod %v. Scaledown processing cannot trigger recovery. "+
+			"Management command: %v, JSON response: %v", scaleDownPodName, wildflyutil.MgmtOpTxnEnableRecoveryListener, jsonResult)
+	}
+	// Enabling the recovery listner may require the server being reloaded
+	isReloadRequired := wildflyutil.ReadJSONDataByIndex(jsonResult["response-headers"], "operation-requires-reload")
+	if isReloadRequiredBool, _ := strconv.ParseBool(isReloadRequired); isReloadRequiredBool {
+		jsonResult, err = wildflyutil.ExecuteMgmtOp(scaleDownPod, jbossHome, wildflyutil.MgmtOpReload)
+		if err != nil {
+			return false, "", fmt.Errorf("Cannot reload application container for scaling down pod %v, error: %v", scaleDownPodName, err)
+		}
+		for serverStateCheckCounter := 0; err != nil && serverStateCheckCounter < reloadRetryCount; serverStateCheckCounter++ {
```

Comment:
Does that mean that the server has 10 seconds to reload?
 My comment at https://github.com/wildfly/wildfly-operator/pull/54#discussion_r311515946 still stands (or maybe you could use a goroutine at least)

Reply:
The server has 10 seconds to reload. After that time it's considered to fail on recovery  processing. The reconcile loop is retried and the server is either already ready to response or it gets another 10 seconds to be reloaded. Maybe this 10 second is not necessary and the reconcile loop just takes care.